be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala

```diff
@@ -459,4 +461,149 @@ object MessageDataProtocol extends DefaultJsonProtocol {
       POPCHA_ADDRESS -> obj.popchaAddress.toJson
     )
   }
+
+  implicit object FederationChallengeFormat extends JsonFormat[FederationChallenge] {
+    final private val PARAM_CHALLENGE_VALUE: String = "value"
+    final private val PARAM_CHALLENGE_VALID_UNTIL: String = "validUntil"
+
+    override def read(json: JsValue): FederationChallenge = json.asJsObject().getFields(PARAM_CHALLENGE_VALUE, PARAM_CHALLENGE_VALID_UNTIL) match {
+      case Seq(value @ JsString(_), validUntil @ JsNumber(_)) => FederationChallenge(
+          value.convertTo[Base16Data],
+          validUntil.convertTo[Timestamp]
+        )
+      case _ => throw new IllegalArgumentException(s"Can't parse json value $json to a FederationChallenge object")
+    }
+
+    override def write(obj: FederationChallenge): JsValue = {
+      var jsObjectContent: ListMap[String, JsValue] = ListMap[String, JsValue](
+        PARAM_OBJECT -> JsString(obj._object.toString),
+        PARAM_ACTION -> JsString(obj.action.toString),
+        PARAM_CHALLENGE_VALUE -> obj.value.toJson,
+        PARAM_CHALLENGE_VALID_UNTIL -> obj.validUntil.toJson
+      )
+      JsObject(jsObjectContent)
+    }
+  }
+
+  implicit object FederationExpectFormat extends JsonFormat[FederationExpect] {
+    final private val PARAM_ID: String = "laoId"
+    final private val PARAM_SERVER_ADDRESS: String = "serverAddress"
+    final private val PARAM_OTHER_ORGANIZER: String = "other_organizer"
+    final private val PARAM_CHALLENGE: String = "challenge"
+
```

Comment:
Same remark as previously. The PARAM should match some keys in the RPC message defined. But once you extracted them, you can use them as you want in your object "FederationExpect". "laoId" -> "lao_id", "serverAddress" -> "server_address", "other_organizer" -> "public_key"